### PR TITLE
feat: draft ModelComparisonReport with dynamic annotation overlap (#209)

### LINKED ISSUE #209: Refactor: Move `plot_venn` into a dedicated `VennComparisonReport` class
### Problem or use case

I want to standardize the visualization logic for model comparisons because currently, the `plot_venn` function resides as a standalone function in `analysis.comparison`. As `refineGEMs` moves toward a more modular, object-oriented reporting architecture, leaving standalone plotting functions scattered across analysis modules creates architectural inconsistency and makes maintenance harder.

### Proposed solution

Create a new `VennComparisonReport` class inside the `classes.reports` module. Migrate the `plot_venn` logic into this new class so that it perfectly aligns with the rest of the updated reporting architecture.

### Affected area

- [x] Python API
- [x] Developer architecture
- [ ] Command-line interface
- [ ] Sphinx / Read the Docs documentation
- [ ] Existing media workflow or media documentation
- [ ] Tests
- [ ] Packaging, installation, or optional dependencies

### Example usage

```text
from refinegems.classes.reports import VennComparisonReport

venn_report = VennComparisonReport(models=[model1, model2, model3])
venn_report.visualise()
```

### Alternatives considered

We considered leaving `plot_venn` in its current location within `analysis.comparison`. While this avoids breaking changes in the short term, it leaves the reporting pipeline fragmented and prevents the reporting classes from being fully unified.

### Compatibility considerations

This will affect existing APIs and workflows. Scripts that currently import `plot_venn` directly from `analysis.comparison` will experience a breaking change. To mitigate this, we can either leave a deprecation wrapper function in the old location that redirects to the new class, or simply note the breaking API change clearly in the documentation and release notes.

### Additional context

This refactoring was explicitly discussed and agreed upon with @cb-Hades during the review and merge of PR #190

### PR DESCRIPTION
**Pull request by dyrpsf**
### Description
Addresses #209

This is a Draft PR to share the initial skeleton and core logic for the new `ModelComparisonReport` class, incorporating the expanded scope discussed in the issue. I am opening this early to get feedback on the overlap algorithm and plotting behavior before finalizing the data export methods.

### Changes Made
*   **Added `ModelComparisonReport`**: Created the new report class inside `src/refinegems/classes/reports.py`.
*   **Overall Annotation Overlap**: Implemented a connected-components (BFS) algorithm in `_calculate_overlap()` that flattens COBRApy annotations. Entities are now matched across models if they share *any* common annotation string, completely ignoring specific namespaces.
*   **Dynamic Plotting Router**: Implemented `visualise()` to automatically route to `venn` (for $\le 4$ models) or `upsetplot` (for $> 4$ models).
*   **Bug Workaround**: In `_plot_upset`, I pre-aggregated the memberships using Python's `Counter` and set `show_counts=False`. This safely bypasses a known `TypeError` crash caused by an incompatibility between `upsetplot` and `pandas >= 2.2.0`.

### Questions for Reviewers
1. Does the connected-components approach for the overall annotation overlap align with what you envisioned for Release 2.0?
2. Are you comfortable leaving `show_counts=False` on the UpSet plots to ensure compatibility with newer pandas versions, or would you prefer a different workaround?

Once the core logic is approved, I will finalize the `to_table()` and `save()` methods.

### CODE REVIEW

**Comment by dyrpsf** [comment]
Thanks @draeger and copilot, I've just pushed a commit to harden the overlap algorithm and fix the edge cases! 

Here is what was updated:
* **Validation:** The class now explicitly requires at least 2 models and rejects duplicate model names.
* **True Namespace Independence:** Stripped the database prefixes during extraction so the algorithm matches the actual ID values, regardless of the database namespace.
* **SBO Term Exclusion:** Ignored SBO terms during the overlap check so broad terms do not accidentally collapse unrelated entities into a single group.
* **Internal ID Fallback:** Fixed the logic to only use the COBRA internal ID as a matching edge if the entity is completely unannotated.
* **Performance:** Swapped the BFS `queue.pop(0)` for `collections.deque.popleft()` to prevent a quadratic bottleneck on models with thousands of entities.

**One open question for @GwennyGit and @draeger :** Copilot noted that the BFS approach creates a "transitive" overlap. For example, if Model A shares an ID with Model B, and Model B shares a *different* ID with Model C, all three will be grouped into one single component. Does this transitive grouping align with your vision for the "overall annotation overlap", or should we restrict it to direct, strict matches only?

**Comment by dyrpsf** [comment]
I updated the logic! Copilot was right about the transitive overlap issue. I swapped out the connected-components BFS algorithm entirely. We now build the memberships directly from the shared annotations, which fixes the $O(k^2)$ performance bottleneck and prevents false overlaps from unannotated entities. I've also implemented `to_table()` and `save()`.

**Comment by dyrpsf** [comment]
I have implemented a Union-Find (Disjoint Set) algorithm in `_calculate_overlap()` to properly group entity nodes by shared annotations, resolving the double-counting issue while remaining performant. I also turned `plot_venn` in `comparison.py` into a deprecation wrapper that routes to the new report.

**Comment by dyrpsf** [comment]
Thanks again, @draeger and Copilot! I reverted the internal logic of `plot_venn` to its original ID-based intersection method to preserve strict backward compatibility for the `rename` dictionary mapping and unannotated entities. The `DeprecationWarning` is still in place to direct users to the new `ModelComparisonReport` for future use. 

This should clear all remaining blockers!

**Comment by dyrpsf** [comment]
Alright, fifth time's the charm! 🚀

1. **Canonical ID Matching:** I resolved both the transitive grouping and the double-counting issues by assigning a strict "canonical ID" (the lexicographically first annotation) to each entity. 
2. **True Delegation:** `ModelComparisonReport` now supports a `match_by="id"` flag. I updated the `plot_venn` wrapper to parse the legacy dictionary format and fully delegate the rendering to the new report using this flag, completely uniting the architecture.
3. **Edge Cases:** Invalid entity types now throw a strict `ValueError`, and missing models are forcefully injected into the `Counter` so `upsetplot`'s index never crashes.

Ready for a final look! Hope so!

**Comment by dyrpsf** [comment]
The canonical ID approach failed the transitive matching test for edge cases like `{a, shared}` and `{b, shared}`. 

To fix this, I have reverted to a **Union-Find (Disjoint Set)** algorithm, but this time I mapped it over the *entity instances* themselves (using an integer index) rather than the annotations. This correctly preserves the "any-shared-annotation" graph connectivity while mathematically eliminating the double-counting issue.

I also addressed the minor edge cases:
- Added strict validation for `match_by` to reject invalid strings.
- Replaced the string-based sentinel with strict integer index separation for unannotated entities.
- Updated the plot titles to dynamically switch between "Annotation Overlap" and "ID Overlap" based on the `match_by` mode.

**Comment by GwennyGit** [comment]
@dyrpsf
Thank you for your work so far. 🙂 
The implementation, from what I gathered so far, sounds good. 

However, I do not like the database prefix stripping as the database prefix is part of a valid CURIE, and if the same ID is used in different databases, which should hopefully not be the case, it does not necessarily mean that it is really the same entity lying behind this ID. So the database prefix is actually important, even though we want to compare the models over annotations from various databases. 

Moreover, if the model has no annotation, but the user would like to get the overlap across all provided models, I think it would be better to include the model’s internal namespace IDs/entity IDs for matching as an optional parameter. So that the user can decide whether to fall back to the model’s internal IDs if no annotation is provided, or ignore the respective model for the analysis. The only issue with that is that the namespace/database is usually missing from the model’s internal IDs, so in that case, the user might need to provide the corresponding database to get the correct CURIE for matching. In either case, I would inform the user that the chosen model contains no annotations at all for the requested entity type. Maybe, without falling back to the model’s internal IDs, the number of entities could be counted and thus shown as unique IDs? I do not know if the Venn plot function from the `venn` package supports this, though. So if I have model A with annotations, model B with annotations, and model C has no annotations and model A and model B overlap, the Venn plot would have one circle for model C that is on the side and not overlapping with the circles for model A and model B, while the circles for model A and B overlap partially.

As you asked about transitive overlap, I like your implementation using the Union-Find algorithm, which preserves it. In general, I would keep the transitive approach, as in the database-independent matching approach it is enough to know that there is an overlap in the models irrespective of if model A, model B and model C share only the same database annotation or model A and model B share one and model B and model C share another one as long as it is the same entity entry referred to. From what I’ve seen so far in models, the transitive approach would give you a more complete overlap, as different model approaches use not only different namespaces for the model entities but also are sometimes quite incomplete in the annotations.

So far, I have had no time to look at the code directly. These are only my answers/suggestions based on the messages and Copilot requests above. I will try to look at the code as soon as possible.

File: src/refinegems/analysis/comparison.py
Comment (GwennyGit) [inline src/refinegems/analysis/comparison.py:1]: Please do not remove/replace docstrings entirely. If they are still true, please keep them as they are and only extend them with new information. As far as I can see, the arguments and returns are still the same, which they should be, except that in some cases, also none can be returned. Please also inform the user that, if none is returned, no figure was produced.

File: src/refinegems/classes/reports.py [hunk 1]
```diff
@@ -2198,3 +2200,201 @@ def save(
...
+        import pandas as pd
+        from collections import Counter
+        
+        rows = []
+        for entity_type, memberships in self.overlap_data.items():
+            counts = Counter([tuple(m) for m in memberships])
+            for models, count in counts.items():
+                rows.append({
+                    "Entity": entity_type,
+                    "Models_Shared": " & ".join(models),
```
Comment (GwennyGit) [inline src/refinegems/classes/reports.py]: Why the join with ‘&’? I would simply return the list here or join by ','.

File: src/refinegems/classes/reports.py
Comment (GwennyGit) [inline src/refinegems/classes/reports.py:1]: Please, change the following:

- [ ] Naming and implementation of ModelComparisonReport
I am wondering about the report type. With the name ModelComparisonReport, I would assume I get a report visualisation showing the overlapping parts of the model. So, like, for the ModelInfoReport, where multiple graphics show different model information, the ModelComparisonReport would give me a graphic that shows one plot for the genes, one for the metabolites, one for the reactions, and, if possible, one for the pathways. As for the current implementation, that would be more like a SingleModelEntityComparisonReport, ModelEntityComparisonReport, or EntityComparisonReport to me.
- [ ] Handling of overlap based on one specific namespace
As far as I can see, the match_by logic currently doesn’t handle the case where a user wants to compare model overlap based on a specific database namespace. The logic for IDs and based on all available annotations seems fine so far, but if I wanted to compare models solely based on annotations from, e.g. the BiGG namespace, I could not do so.
- [ ] Using getter and setters for the class attributes
I would recommend using getters and setters for class attributes with specific requirements, such as `model_names`.
- [ ] Handling of empty model names
In some cases, models may not have an internal model ID, which can cause issues if no rename list is provided. -> This could also be done with a getter and setter.
- [ ] Not all entities covered
The entities covered should include not only genes, metabolites, and reactions but also pathways, which can be stored in a model with the Group plugin. However, pathways are sadly a special case: some models include them with the Group plugin, some include them solely as annotations within the reaction annotations, and some do not include them at all. Moreover, the relationship is not always specified correctly, as some models state BQB_IS also for pathway annotations within the reaction annotations. Within refineGEMs, we have functionality to add Pathway information via the Group Plugin, and at the same time, the pathway annotation is added as BQB_OCCURS_IN to the reaction annotations.
- [ ] `entity_type` as class attribute
I would store the requested entity type as a class attribute, as it is needed for every function. -> Requires less user input.

**Review by GwennyGit — CHANGES_REQUESTED** [review]
Please see the points in the comments to the specific files for changes.

**Comment by GwennyGit** [comment]
Regarding the two questions you stated above:

> 
    1. Does the connected-components approach for the overall annotation overlap align with what you envisioned for Release 2.0?
    2. Are you comfortable leaving show_counts=False on the UpSet plots to ensure compatibility with newer pandas versions, or would you prefer a different workaround?

1. I think number one should be answered already by my comments on the code.
2. I do not like the idea to hardcode such a setting. Could you please elaborate more on this issue?

File: src/refinegems/analysis/comparison.py
Comment (dyrpsf) [reply src/refinegems/analysis/comparison.py:1]: Done! I've restored the original docstrings alongside the new deprecation warning. I also updated the return description to explicitly note that None is returned when no figure is produced.

File: src/refinegems/classes/reports.py (hunk 1 above)
Comment (dyrpsf) [reply src/refinegems/classes/reports.py]: Good catch. I've updated the formatting in `to_table()` to join the model names with a comma (`", "`) for better readability.

File: src/refinegems/classes/reports.py
Comment (dyrpsf) [reply src/refinegems/classes/reports.py:1]: I completely agree with these structural changes. I've renamed the class to `EntityComparisonReport`, made `entity_type` a class attribute, implemented `@property` getters/setters for `model_names` (with a fallback for missing internal IDs), added pathway support, and updated `match_by` to filter by specific namespaces. These are all live in the latest commit!

**Comment by dyrpsf** [comment]
@GwennyGit Thank you for the thorough review! I agree with your architectural points. I have just pushed a refactor that renames the class to `EntityComparisonReport`, makes `entity_type` a class attribute, adds getters/setters for model names, restores the database prefixes, adds specific namespace filtering to `match_by`, and includes pathway support.

To answer your question regarding the `show_counts=False` hardcoding:
There is a known bug between the `upsetplot` package and `pandas >= 2.2.0`. When `upsetplot` tries to draw the count numbers above the bars, newer versions of pandas accidentally return an array instead of a scalar for the coordinate, causing Matplotlib to instantly crash with a `TypeError`. Disabling `show_counts` is currently the standard workaround to prevent the report from crashing on users' machines who have updated pandas. Let me know if you are okay keeping this workaround, or if you'd prefer I implement a version check (e.g., `if pd.__version__ >= '2.2.0'`) to only disable it for affected versions!

File: src/refinegems/classes/reports.py
Comment (GwennyGit) [inline src/refinegems/classes/reports.py:1]: Thank you for adjusting the code. 
The changes look good so far. I would add minor changes before merging into the dev-2.1 branch, but I will do that myself. 🙂

If you like, you’re also welcome to add a class, ModelComparisonReport/MultiEntityComparisonReport, that uses EntityComparisonReport to generate reports for multiple user-specified entities. I think that would be great from a user perspective. Maybe with the option to get either a single plot containing all plots, or all single plots.

**Comment by GwennyGit** [comment]
Regarding the `show_counts` issue, I think it would be best to use the version check, and if the version of pandas is 'wrong', use a self-implemented version so the user basically sees the same outcome. I think that should work after the figure is already created with `upsetplot`. @dyrpsf What do you think? 🤔

Moreover, I looked more closely at the [issue](https://github.com/jnothman/UpSetPlot/issues/301), and it does not seem to come from `pandas` but rather from `numpy`. So the question would also be more if we ask for the pandas or the numpy version. There is also a fork of `upsetplot` that addresses the issue, but it needs testing and lacks versioning, which is kind of bad. I think for either solution, at least an `@NOTE` or `@WARNING` should be added to the code, with the reasoning for the changes, for later maintainers/developers.

**Comment by dyrpsf** [comment]
@GwennyGit Good catch! You are entirely correct - I read through the linked issue, and the strict scalar casting failure was indeed introduced in NumPy 2.4.0. 

I've just pushed a commit that imports `numpy` inside `_plot_upset`, parses `np.__version__`, and dynamically sets `show_counts=True` for anyone on `< 2.4.0`, while safely disabling it for `>= 2.4.0`. 

I also added a detailed `@WARNING` comment right above the `UpSet` instantiation explaining the NumPy 2.4 crash and linking to `jnothman/UpSetPlot#301` so future developers know exactly why the version check is there.

Since you mentioned you'd take care of the final minor changes before merging, I'll leave the rest in your hands. Thank you so much for the incredibly helpful review process!
